storage/s3.go

```diff
@@ -554,14 +581,58 @@ func (s *S3) Put(
 		input.ContentEncoding = aws.String(contentEncoding)
 	}
 
+	// add retry code to the object metadata
+	if s.noSuchUploadRetryCount > 0 {
+		input.Metadata[META_DATA_S5CMD_RETRY_CODE_KEY] = generateRetryCode()
+	}
+
 	_, err := s.uploader.UploadWithContext(ctx, input, func(u *s3manager.Uploader) {
 		u.PartSize = partSize
 		u.Concurrency = concurrency
 	})
 
+	attempts := 0
+	for ; attempts < s.noSuchUploadRetryCount && err != nil; attempts++ {
+		// check if the error is NoSuchUpload error
+		if !strings.Contains(err.Error(), s3.ErrCodeNoSuchUpload) {
+			break
+		}
+		// check if object exists in the destination
+		obj, sErr := s.Stat(ctx, to)
+		if sErr == nil {
+			// if object has the retry code we provided then it means
+			// the upload was succesfull despite the received error.
+			if obj.retryCode == retryCode {
+				err = nil
+				break
+			}
+		}
+
+		msg := log.DebugMessage{Err: fmt.Sprintf("Retrying to upload %v upon error: %q", to, err.Error())}
+		log.Debug(msg)
+
+		// renew retry code
+		input.Metadata[META_DATA_S5CMD_RETRY_CODE_KEY] = generateRetryCode()
+
+		_, err = s.uploader.UploadWithContext(ctx, input, func(u *s3manager.Uploader) {
+			u.PartSize = partSize
+			u.Concurrency = concurrency
+		})
+	}
+	if err != nil && strings.Contains(err.Error(), s3.ErrCodeNoSuchUpload) && s.noSuchUploadRetryCount > 0 {
+		err = awserr.New(s3.ErrCodeNoSuchUpload, fmt.Sprintf("RetryOnNoSuchUpload: %v attempts to retry resulted in %v", attempts, s3.ErrCodeNoSuchUpload), err)
+		log.Error(log.ErrorMessage{Err: err.Error()})
+	}
+
 	return err
 }
```

Comment:
Instead of checking then breaking loop depending on error, we might check error and retry count first then return early if these conditions are not met. It would make it more readable IMHO.

Reply:
If it returns early depending on decision to retry, moving retry logic to another function/method would makes this function less crowded.

I was thinking something like;

```go
_, err := s.uploader.UploadWithContext(/* */)

if ! /* conditions to retry /* {
	return err
}

return s.someRetryMethod()


```

Reply:
Could you give context on why retry code is re-generated on each retry? Could not the same one re-used?

Reply:
> moving retry logic to another function/method

I'd some concerns about it. First, I think we should limit the AWS-SDK calls to the storage methods,  so having another method that also uses AWS-SDK (`s.uploader.UploadWithContext`) might be an extra maintenance work. Second, that method needs (in fact [needed](https://github.com/Kucukaslan/s5cmd/blob/9770a80d009bb7ad29588e80b6e81caf374d8b6a/storage/s3.go#L593)) a lot of (5+) parameters. I would prefer not to move a logic that needs so many arguments. 

Nevertheless, I agree that readability is improved by moving it to another method. Especially after considering that that logic was an exceptional/tangential case.

> why retry code is re-generated on each retry
I thought that, at each turn (in the loop) we would check if the _the previous_ upload was successful. However, after discussion, I come to think that if _any of the previous_ uploads were successful than we should not retry the upload. So, I'll use the same retryCode for all of them.